Modules/FindCUDA/select_compute_arch.cmake

```diff
@@ -0,0 +1,168 @@
+    endif()
+  endif()
+
+  if(NOT CUDA_GPU_DETECT_OUTPUT)
+    message(STATUS "Automatic GPU detection failed. Building for all known architectures.")
+    set(${OUT_VARIABLE} ${CUDA_KNOWN_GPU_ARCHITECTURES} PARENT_SCOPE)
+  else()
+    set(${OUT_VARIABLE} ${CUDA_GPU_DETECT_OUTPUT} PARENT_SCOPE)
+  endif()
+endfunction()
+
+
+################################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_NAME
+# Usage:
+#   SELECT_NVCC_ARCH_FLAGS(out_variable)
+function(CUDA_SELECT_NVCC_ARCH_FLAGS out_variable)
+  # List of arch names
+# CUDA_ARCH_NAME: One of:
+  set(archs_names "Fermi" "Kepler" "Kepler+Tegra" "Kepler+Tesla" "Maxwell" "Maxwell+Tegra" "Pascal" "All" "Manual")
+  set(archs_name_default "All")
+  if(NOT CMAKE_CROSSCOMPILING)
+    list(APPEND archs_names "Auto")
+    set(archs_name_default "Auto")
+  endif()
+
+  # SET CUDA_ARCH_NAME strings (so it will be seen as dropbox in CMake-Gui)
+  set(CUDA_ARCH_NAME ${archs_name_default} CACHE STRING "Select target NVIDIA GPU achitecture.")
+  set_property( CACHE CUDA_ARCH_NAME PROPERTY STRINGS "" ${archs_names} )
+  mark_as_advanced(CUDA_ARCH_NAME)
+
+  # verify CUDA_ARCH_NAME value
+  list(FIND archs_names ${CUDA_ARCH_NAME} arch_found)
+
+  if(${arch_found} EQUAL -1)
+    message(FATAL_ERROR "Only ${archs_names} architecture names are supported.")
+  endif()
+
+  if(${CUDA_ARCH_NAME} STREQUAL "Manual")
+    set(CUDA_ARCH_BIN ${CUDA_KNOWN_GPU_ARCHITECTURES} CACHE STRING "SpecIFy 'real' GPU architectures to build binaries for, BIN(PTX) format is supported")
+    set(CUDA_ARCH_PTX "50"                     CACHE STRING "SpecIFy 'virtual' PTX architectures to build PTX intermediate code for")
+    mark_as_advanced(CUDA_ARCH_BIN CUDA_ARCH_PTX)
+  else()
+    unset(CUDA_ARCH_BIN CACHE)
+    unset(CUDA_ARCH_PTX CACHE)
+  endif()
+
+  # Allow a user to specify architecture from env
+  if($ENV{CUDA_ARCH_BIN})
+    set(CUDA_ARCH_NAME "Manual")
+    set(CUDA_ARCH_BIN $ENV{CUDA_ARCH_BIN})
+    unset(CUDA_ARCH_PTX)
+  endif()
+
+  if(${CUDA_ARCH_NAME} STREQUAL "Fermi")
+    set(cuda_arch_bin "2.0 2.1(2.0)")
+  elseif(${CUDA_ARCH_NAME} STREQUAL "Kepler+Tegra")
+    set(cuda_arch_bin "3.2 3.2")
+  elseif(${CUDA_ARCH_NAME} STREQUAL "Kepler+Tesla")
+    set(cuda_arch_bin "3.7 3.7")
+  elseif(${CUDA_ARCH_NAME} STREQUAL "Kepler")
+    set(cuda_arch_bin "3.0 3.5")
+  elseif(${CUDA_ARCH_NAME} STREQUAL "MAxwell+Tegra")
+    set(cuda_arch_bin "5.3 5.3")
+  elseif(${CUDA_ARCH_NAME} STREQUAL "Maxwell")
+    set(cuda_arch_bin "5.0 5.2")
+  elseif(${CUDA_ARCH_NAME} STREQUAL "Pascal")
+    set(cuda_arch_bin "6.0 6.2")
+  elseif(${CUDA_ARCH_NAME} STREQUAL "All")
+    set(cuda_arch_bin ${CUDA_KNOWN_GPU_ARCHITECTURES})
+  elseif(${CUDA_ARCH_NAME} STREQUAL "Auto")
+    CUDA_DETECT_INSTALLED_GPUS(cuda_arch_bin)
+  else()  # (${CUDA_ARCH_NAME} STREQUAL "Manual")
```

Comment:
It might be useful to verify CUDA_ARCH_NAME == "Manual" or put in a default case with an error message.

```
elseif(${CUDA_ARCH_NAME} STREQUAL "Manual")
  set(cuda_arch_bin ${CUDA_ARCH_BIN})
else()
  message(SEND_ERROR "Unknown CUDA_ARCH_NAME (${CUDA_ARCH_NAME})")
endif()
```